libs/utils/src/lsn.rs

```diff
@@ -13,10 +13,77 @@ use crate::seqwait::MonotonicCounter;
 pub const XLOG_BLCKSZ: u32 = 8192;
 
 /// A Postgres LSN (Log Sequence Number), also known as an XLogRecPtr
-#[derive(Clone, Copy, Eq, Ord, PartialEq, PartialOrd, Hash, Serialize, Deserialize)]
-#[serde(transparent)]
+#[derive(Clone, Copy, Eq, Ord, PartialEq, PartialOrd, Hash)]
 pub struct Lsn(pub u64);
 
+impl Serialize for Lsn {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        if serializer.is_human_readable() {
+            format!("{:X}/{:X}", self.0 >> 32, self.0 & 0xffffffff).serialize(serializer)
+        } else {
+            self.0.serialize(serializer)
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for Lsn {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        struct LsnVisitor;
+
+        impl<'de> Visitor<'de> for LsnVisitor {
+            type Value = Lsn;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                formatter.write_str("u64 value in either bincode form(u64) or serde_json form({upper_u32_hex}/{lower_u32_hex})")
+            }
+
+            fn visit_u64<E>(self, v: u64) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                Ok(Lsn(v))
+            }
+
+            fn visit_string<E>(self, v: String) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                self.visit_str(&v)
+            }
+
+            fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                let mut splitter = v.trim().split('/');
+                if let (Some(left), Some(right), None) =
+                    (splitter.next(), splitter.next(), splitter.next())
+                {
+                    let left_num =
+                        u32::from_str_radix(left, 16).map_err(|_| E::custom(LsnParseError))?;
+                    let right_num =
+                        u32::from_str_radix(right, 16).map_err(|_| E::custom(LsnParseError))?;
+                    Ok(Lsn((left_num as u64) << 32 | right_num as u64))
+                } else {
+                    Err(E::custom(LsnParseError))
+                }
```

Comment:
This could use Lsn::from_str

Reply:
Done https://github.com/neondatabase/neon/pull/5335/commits/8251563491214e845bff4343e16958e18b6d59f1